FE scale audit: Implement backend paging in org membership view

philipp-spiess opened the pull request:
Closes #43417 

As part of the [FE scale audit](https://docs.google.com/spreadsheets/d/1tqpi22csQJNlk7Q9JFWcFj09VbWOiFdp-Vf18qJsWHc/edit#gid=0) I've identified a few areas that lack basic server-side pagination. In this case, the org members page would not paginate at all leading to a long list if you have many members.

This is a simple fix that adds pagination based on LIMIT/OFFSET to the resolver (limit/offset was already implemented, I only had to generate the GraphQL pagination tokens correctly). I've also decided to add a search query endpoint to the members list so that it's easy to e.g. revoke access to a single user even if you have hundreds of users in the org. 

<img width="1115" alt="Screenshot 2022-10-25 at 14 42 08" src="https://user-images.githubusercontent.com/458591/197775799-af951256-de62-47ba-b506-4ed70b5dda7d.png">

To make testing easier, I've also added support for the new `after` param to the `users` root GraphQL query. This does not break existing APIs though as the field is optional and the previous "let's increase the limit with every page" logic still works.

## Test plan

- I did extensive testing in the GraphQL console
- I added a test case for a basic pagination example to the users test
- https://www.loom.com/share/ebe9da1a93ba436896705dc973b5b932

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.16 kb) | 0.00% (+0.16 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/org-membership-pagination/compare-report.html;tab=live_object) for a full comparison between the commits d8e086c and 62fe752 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

philipp-spiess commented on client/web/src/org/members/OrgMembersListPage.tsx:
```diff
@@ -205,7 +205,7 @@ export const OrgMembersListPage: React.FunctionComponent<React.PropsWithChildren
     const { data, loading, error, refetch } = useQuery<OrganizationMembersResult, OrganizationMembersVariables>(
         ORG_MEMBERS_QUERY,
         {
-            variables: { id: org.id },
```
Revert and create a follow up issue to remove the "new" org member list page that is gated by `newMembersInviteEnabled`

philipp-spiess commented on client/web/src/org/settings/members-v1/OrgSettingsMembersPage.tsx:
```diff
@@ -252,11 +253,11 @@ export class OrgSettingsMembersPage extends React.PureComponent<Props, State> {
                         updates={this.userUpdates}
                         history={this.props.history}
                         location={this.props.location}
+                        defaultFirst={2}
```
rm

philipp-spiess commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -140,28 +141,43 @@ func (o *OrgResolver) SettingsURL() *string { return strptr(o.URL() + "/settings
-		users[i] = user
+		limitOffset.Offset = int(cursor)
 	}
-	return &staticUserConnectionResolver{db: o.db, users: users}, nil
```
I decided to use the already implemented `userConnectionResolver` here instead of going with a custom one. Not quite sure if I like the result because the security checks are a bit awkward now, so if you think we should have a custom resolver, let me know.

The benefit of course is that we now also have tokenized scrolling support for the  `users` root query.

mrnugget commented on cmd/frontend/graphqlbackend/org.go:
```diff
@@ -140,28 +141,43 @@ func (o *OrgResolver) SettingsURL() *string { return strptr(o.URL() + "/settings
 func (o *OrgResolver) CreatedAt() gqlutil.DateTime { return gqlutil.DateTime{Time: o.org.CreatedAt} }
 
-func (o *OrgResolver) Members(ctx context.Context) (*staticUserConnectionResolver, error) {
-	// 🚨 SECURITY: Only org members can list other org members.
```
I'd keep the check up here, in `Members`, instead of moving it down to `Nodes`. Two reasons: you have duplicated code in `Count()` and `Nodes()` (not that bad) but you also don't have the check in `PageInfo`.

philipp-spiess replied:
I was worried about two things:

- Right now, the userConnectionResolver requires you to be site admin to list users. Since I've reused it here, I'll need to change it so that even non-site-admins can list users. I can allowlist the query when a valid `OrgId` is set, but that would split security checks between two places.
- I want to avoid possible security bypasses if another caller would start passing an org id to the users resolver without having security checks for that set up.
 
I guess what I could do is also move the security check that was already inside the `userConnectionResolver` up to the callsite? It's only one that is not checked right now. That would make it clearer that security checks belong into the callsits.

mrnugget replied:
That makes sense then (forgot about that the resolver is used in multiple places). Still: how about abstracting the two calls into a separate func (`func checkMembersAccess(ctx, orgID int32) error`) and possibly adding it to the top entrypoint where possible?

mrnugget commented on internal/database/users.go:
```diff
@@ -802,6 +802,8 @@ type UsersListOptions struct {
 	// UserIDs specifies a list of user IDs to include.
 	UserIDs []int32
+	// Only show users inside this org
+	OrgId *int32
```
I think it's better to make this an `int32`. We generally use `!= 0` checks in the database layer and its option structs. If you don't use a pointer here, the code in the graphql layer also becomes easier.

```suggestion
	OrgID int32
```

mrnugget commented on internal/database/users.go:
```diff
@@ -890,6 +892,17 @@ func (*userStore) listSQL(opt UsersListOptions) (conds []*sqlf.Query) {
+	WHERE
+		org_members.user_id = u.id
+		AND org_members.org_id = %d
+		AND org_members.org_id <> 0)
```
We have a foreign key constraint on `org_members.org_id`, so technically it can't ever be 0. No need to check here.

mrnugget commented on internal/database/users.go:
```diff
@@ -890,6 +892,17 @@ func (*userStore) listSQL(opt UsersListOptions) (conds []*sqlf.Query) {
 		}
 	}
+	if opt.OrgId != nil {
+		conds = append(conds, sqlf.Sprintf(`
```
Can you pull this out, like `listUsersInactiveCond` above?

mrnugget commented on cmd/frontend/graphqlbackend/users_test.go:
```diff
@@ -52,6 +52,91 @@ func TestUsers(t *testing.T) {
+func TestUsers_Pagination(t *testing.T) {
+	users := database.NewMockUserStore()
+	users.GetByCurrentAuthUserFunc.SetDefaultReturn(&types.User{SiteAdmin: true}, nil)
+	users.ListFunc.SetDefaultReturn([]*types.User{
+		{Username: "user1"},
+		{Username: "user2"},
+	}, nil)
```
Instead of `SetDefaultReturn` I'd rather set a custom func here that checks for the opts and then returns a dynamic list of users.

Leaving the test as it is right now, I could comment out 80% of your pagination code in the GraphQL layer and it would still pass, because of the two mocked out functions here.

Instead I want a test here that makes sure the pagination actually passes the correct values on to the database.

_Ideally_ we'd also have an integration-y test that creates users in the database and doesn't mock the DB out and confirms that it works.

mrnugget commented on cmd/frontend/graphqlbackend/users_test.go:
```diff
@@ -52,6 +52,91 @@ func TestUsers(t *testing.T) {
+		{Username: "user4"},
+	}, nil)
+
+	RunTests(t, []*Test{
```
If you change the approach as I describe in previous comment, you can actually use `RunTests` (plural) and put both test cases in here. Otherwise I suggest you use `RunTest`.

mrnugget commented on cmd/frontend/graphqlbackend/users.go:
```diff
@@ -138,34 +167,43 @@ func (r *userConnectionResolver) TotalCount(ctx context.Context) (int32, error)
-}
 
-func (r *staticUserConnectionResolver) TotalCount() int32 { return int32(len(r.users)) }
+	// In case the number of results happens to be the same as the limit,
+	// we need another query to get accurate total count with same cursor
+	// to determine if there are more results than the limit we set.
+	totalCount, err := r.TotalCount(ctx)
```
You could also change the calls to `db.users().List()` to always query 1 more and then cut that off from the results.

mrnugget commented on cmd/frontend/graphqlbackend/users.go:
```diff
@@ -122,9 +141,19 @@ func (r *userConnectionResolver) Nodes(ctx context.Context) ([]*UserResolver, er
+				return 0, errors.New("must be a member of this organization to view members")
+			}
+			return 0, err
+		}
```
See comment above. I don't think we should duplicate these checks again in two different methods

mrnugget commented on cmd/frontend/graphqlbackend/users.go:
```diff
@@ -138,34 +167,43 @@ func (r *userConnectionResolver) TotalCount(ctx context.Context) (int32, error)
+	if r.useCache() {
+		users, _, err = r.compute(ctx)
+	} else {
+		users, err = r.db.Users().List(ctx, &r.opt)
```
Is this really required, to fetch the whole list of users _again_ just to compute the `PageInfo`?

philipp-spiess replied:
We somehow need to know how many items were loaded, hmm.

Other implementations of this were doing the same thing but `r.compute()` would have cached the result, I think this does not work here because we have this uncached case.

I think I'll need to understand why we have this weird uncached codepath

mrnugget replied:
Yeah, that uncache/cached thing is the problem :/

philipp-spiess replied:
Okay I looked into it and I think we can remove the `useCache` branching here. With the query gaining the pageInfo feature, it's clear that no matter the `activePeriod`, the caching would always be useful. This is, accidentally, also a nice code cleanup.

mrnugget reviewed (commented):
Left a few comments, most important ones are the one on the test and the pointer for `OrgId`

philipp-spiess commented:
@mrnugget This is ready for another review 😎

mrnugget reviewed (approved):
Looks good!

I'd really love to see a higher-level integration test next to the `TestUsers_Pagination` test though that exercises also the database layer. Doesn't have to be big, but just makes sure that it works as intended with more parts involved.

See this test here for an example of what I mean: https://github.com/sourcegraph/sourcegraph/blob/13bc99ac6b3504a9fa2b5507c967a44ae20c945b/cmd/frontend/graphqlbackend/repositories_test.go#L524

Filtering/pagination has a tricky database/GraphQL interplay and I added that test to make sure it actually works as it should. That test throws some repositories in the database and then queries them.

For this here, I'd create an org with ~3 members (like [this here](https://github.com/sourcegraph/sourcegraph/blob/13bc99ac6b3504a9fa2b5507c967a44ae20c945b/internal/database/orgs_test.go#L254-L291)) and then make sure `first: 1`, `first: 2`, `first: 3`, `first: 4` work as intended. Also with different contexts to make sure auth works.

philipp-spiess commented:
@mrnugget Sorry for the long wait here (I was on PTO): Fixed up the issues and added an integration tests to the users query. Should we add one for the org memberships query as well (This can be done as a follow-up though!).
